dipy/denoise/localpca.py

```diff
@@ -0,0 +1,137 @@
+
+        #     sigma = (np.ones(arr.shape, dtype=np.float64) * sigma[..., np.newaxis])
+        #     tou = (np.ones(arr.shape, dtype=np.float64) * tou[..., np.newaxis])
+        # else:
+        #     sigma = np.ones(arr.shape, dtype=np.float64) * sigma
+        #     tou = np.ones(arr.shape, dtype=np.float64) * tou    
+        
+        # loop around and find the 3D patch for each direction at each pixel
+        
+        # declare arrays for theta and thetax
+        theta = np.zeros(arr.shape,dtype = np.float64)
+        thetax = np.zeros(arr.shape,dtype = np.float64)
+
+        patch_size = 2 * patch_radius + 1
+
+        for k in range(patch_radius, arr.shape[2] - patch_radius, 1):
+            print(k)
+            for j in range(patch_radius, arr.shape[1] - patch_radius, 1):
+                for i in range(patch_radius, arr.shape[0] - patch_radius , 1):
+                    
+                    X = np.zeros((patch_size * patch_size * patch_size, arr.shape[3]))
+                    M = np.zeros(arr.shape[3])
+
+                    temp = arr[i - patch_radius : i + patch_radius + 1,j - patch_radius : j + patch_radius + 1,
+                         k - patch_radius : k + patch_radius + 1,:]
+                    X = temp.reshape(patch_size * patch_size * patch_size, arr.shape[3])
+                    # compute the mean and normalize
+                    M = np.mean(X,axis=1)
+                    X = X - np.array([M,]*X.shape[1],dtype=np.float64).transpose()
+                        
+                    # using the PCA trick
+                    # Compute the covariance matrix C = X_transpose X
+                    C = np.transpose(X).dot(X)
+                    C = C/arr.shape[3]
+                    # compute EVD of the covariance matrix of X get the matrices W and D, hence get matrix Y = XW
+                    # Threshold matrix D and then compute X_est = YW_transpose D_est
+                    [d,W] = np.linalg.eigh(C)
+                    # for sigma estimate we perform the median estimation
+
+                    # find the median of the standard deviation of the eigenvalues
+                    median_sqrt = np.median(np.sqrt(d[d > alpha]))
+                    if(np.isnan(median_sqrt)):
+                        median_sqrt = 0
+                    # Chop of the positive eigenvalues whose standard deviation is more that 2 times that of the above quantity
+                    # Take the remaining eigenvalues and estimate sigma
+                    sigma[i,j,k] = beta * beta * np.median(d[np.sqrt(d[d > alpha]) < 2 * median_sqrt])
+                    if(np.isnan(sigma[i,j,k])):
+                        sigma[i,j,k] = 0
+
+                    d[d < sigma[i,j,k]] = 0
+                    D_hat = np.diag(d)
+                    Y = X.dot(W)
+                    # When the block covers each pixel identify it into the label matrix theta
+                    X_est = Y.dot(np.transpose(W))
+                    X_est = X_est.dot(D_hat)
```

Comment:
Perhaps a good test will be comparing these lines of code with: 
X_est = X.dot(D_hat)  # your previous way to compute X_est